### DISCUSSION TITLE
unpredictability of csrf token in verification link response

### DISCUSSION THREAD

**Discussion by ujaved-doc-ai** [ory/kratos#2238] [Q&A] [answered]
I'm implementing an api [verification flow](https://www.ory.sh/kratos/docs/self-service/flows/verify-email-account-activation). The behavior I'm observing is that when the user clicks on the verification link sent to their email, it works sometimes but often doesn't. Meaning that the user isn't verified even after clicking the link, and the kratos error is: `The verification token is invalid or has already been used. Please retry the flow.` This is true even if the link is clicked just once and before it expires.

I have tracked this down to the cases where the kratos response on clicking the verification link contains a `csrf_token` as a `set-cookie` header. For example:

```
> GET /public/self-service/verification?flow=c5b7e9ee-d442-45f1-9835-38ba24e8b633&token=7WJCnaRQc07zQUpXS3vwUnft2pu3tuCs HTTP/1.1
> Host: smartomix-sso-beta.doc.ai
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 303 See Other
< cache-control: private, no-cache, no-store, must-revalidate
< content-type: text/html; charset=utf-8
< location: https://smartomix-beta.doc.ai/emailVerification/success?flow=4dd40395-8b71-4a11-a840-0a600039671c
< set-cookie: csrf_token_1293c81ee8f8c1a915ade759cb7f10ad36f707d4ef634fb6c5c09d370239a099=hSWPY6G/7lpH6ZG89b6ikZLEImHVOJQCNrYVXrEG2Ds=; Path=/; Max-Age=31536000; HttpOnly; Secure; SameSite=Lax
< vary: Cookie
< date: Fri, 18 Feb 2022 23:12:52 GMT
< content-length: 124
< x-envoy-upstream-service-time: 22
< server: envoy
```

Sometimes the csrf_token is absent and verification succeeds with Kratos log: `Post verification execution hooks completed successfully.` This difference seems to be non-deterministic. Why is this the case?  

Also, for the case when the csrf token is present in the response header, whats the way to complete verification?

Thanks.

**Comment by aeneasr** [comment, marked as answer]
Hello, are you 100% confident that the address is not verified? The error you show means two things:

- the link was used and the user is verified
- the link expired

You might also be hitting: https://github.com/ory/kratos/issues/1451#issue-926006367

**Reply by vinckr** [reply]
Hey @ujaved-doc-ai , just checking in if your problem still persists. 
I will mark this as answered for now, but feel free to come back to the discussion any time.
Thanks a ton :bee:
